features/mbedtls/targets/TARGET_NUVOTON/TARGET_NUC472/aes/aes_alt.c

```diff
@@ -21,304 +21,220 @@
+ *       2) <= MAX_DMA_CHAIN_SIZE
+ */
 static void __nvt_aes_crypt( mbedtls_aes_context *ctx,
-                          const unsigned char input[16],
-                          unsigned char output[16], int dataSize)
+                             const unsigned char *input,
+                             unsigned char *output, size_t dataSize)
 {
-		unsigned char* pIn;
-	  unsigned char* pOut;
-
-//	  mbedtls_trace("=== %s \r\n", __FUNCTION__);
-	  dumpHex(input,16);
- 
-    AES_Open(ctx->channel, ctx->encDec, ctx->opMode, ctx->keySize, ctx->swapType);
-    AES_SetInitVect(ctx->channel, ctx->iv);
-		if( ((uint32_t)input) & 0x03 )
-		{
-			memcpy(au8InputData, input, dataSize);
-			pIn = au8InputData;
-		}else{
-		  pIn = (unsigned char*)input;
+    const unsigned char* pIn;
+    unsigned char* pOut;
+
+    MBED_ASSERT((dataSize % 16 == 0) && (dataSize <= MAX_DMA_CHAIN_SIZE));
+    
+    /* AES DMA buffer has the following requirements:
+     * (1) Word-aligned buffer base address
+     * (2) 16-byte aligned buffer size
+     * (3) Located in 0x20000000-0x2FFFFFFF region
+     */
+    if ((! crypto_dma_buff_compat(au8OutputData, MAX_DMA_CHAIN_SIZE, 16)) ||
+        (! crypto_dma_buff_compat(au8InputData, MAX_DMA_CHAIN_SIZE, 16))) {
+        error("Buffer for AES alter. DMA requires to be word-aligned and located in 0x20000000-0x2FFFFFFF region.");
     }
-		if( (((uint32_t)output) & 0x03) || (dataSize%4))   // HW CFB output byte count must be multiple of word
-		{		
-			pOut = au8OutputData;
-		} else {
-		  pOut = output;
-    }			
-
-    AES_SetDMATransfer(ctx->channel, (uint32_t)pIn, (uint32_t)pOut, dataSize);		
 
-    g_AES_done = 0;
-    AES_Start(ctx->channel, CRYPTO_DMA_ONE_SHOT);
-    while (!g_AES_done);
+    /* TODO: Change busy-wait to other means to release CPU */
+    /* Acquire ownership of AES H/W */
+    while (! crypto_aes_acquire());
+    
+    /* Init crypto module */
+    crypto_init();
+    /* Enable AES interrupt */
+    AES_ENABLE_INT();
+    
+    /* We support multiple contexts with context save & restore and so needs just one 
+     * H/W channel. Always use H/W channel #0. */
+
+    /* AES_IN_OUT_SWAP: Let H/W know both input/output data are arranged in little-endian */
+    AES_Open(0, ctx->encDec, ctx->opMode, ctx->keySize, AES_IN_OUT_SWAP);
+    AES_SetInitVect(0, ctx->iv);
+    AES_SetKey(0, ctx->keys, ctx->keySize);
+    /* AES DMA buffer requirements same as above */
+    if (! crypto_dma_buff_compat(input, dataSize, 16)) {
+        memcpy(au8InputData, input, dataSize);
+        pIn = au8InputData;
+    } else {
+        pIn = input;
+    }
+    /* AES DMA buffer requirements same as above */
+    if (! crypto_dma_buff_compat(output, dataSize, 16)) {
+        pOut = au8OutputData;
+    } else {
+        pOut = output;
+    }
 
-    if( pOut != output ) memcpy(output, au8OutputData, dataSize);
-		dumpHex(output,16);
+    AES_SetDMATransfer(0, (uint32_t)pIn, (uint32_t)pOut, dataSize);
```

Comment:
@ccli8 Does the H/W work with equal input and output buffer pointers?

Reply:
@hanno-arm H/W just has limited support for input/output buffers overlapped. For simplicity, I guard from it in https://github.com/ARMmbed/mbed-os/pull/4925/commits/08c4e823545fdb0de286e1bc14c02282d5accaca.